airbyte-integrations/connectors/destination-keen/src/main/java/io/airbyte/integrations/destination/keen/KeenTimestampService.java

```diff
@@ -0,0 +1,187 @@
+    String streamName = message.getRecord().getStream();
+    CursorField cursorField = streamCursorFields.get(streamName);
+    JsonNode data = message.getRecord().getData();
+    if (timestampInferenceEnabled && cursorField != null) {
+      try {
+        String timestamp = parseTimestamp(cursorField, data);
+        injectTimestamp(data, timestamp);
+      } catch (Exception e) {
+        // If parsing of timestamp has failed, remove stream from timestamp-parsable stream map,
+        // so it won't be parsed for future messages.
+        LOGGER.info("Unable to parse timestamp field: " + cursorField.name);
+        streamCursorFields.remove(streamName);
+        injectTimestamp(data, Instant.ofEpochMilli(message.getRecord().getEmittedAt()).toString());
+      }
+    }
+    else {
+      injectTimestamp(data, Instant.ofEpochMilli(message.getRecord().getEmittedAt()).toString());
+    }
+    return data;
+  }
+
+  private void injectTimestamp(JsonNode data, String timestamp) {
+    ObjectNode root = ((ObjectNode) data);
+    root.set("keen", JsonNodeFactory.instance.objectNode().put("timestamp", timestamp));
+  }
+
+  private String parseTimestamp(CursorField cursorField, JsonNode data) {
+    return switch (cursorField.type) {
+      case NUMBER -> dateFromNumber(data.get(cursorField.name).asLong());
+      case STRING -> parser
+          .parse(data.get(cursorField.name).asText())
+          .get(0).getDates()
+          .get(0)
+          .toInstant()
+          .toString();
+      default -> throw new IllegalStateException("Unexpected value: " + cursorField.type);
+    };
+  }
+
+  private String dateFromNumber(Long timestamp) {
+    // if cursor value is below given threshold, assume that it's not epoch timestamp but ordered id
+    if (timestamp < SECONDS_FROM_EPOCH_THRESHOLD) {
+      throw new IllegalArgumentException("Number cursor field below threshold: " + timestamp);
+    }
+    // if cursor value is above given threshold, then assume that it's Unix timestamp in milliseconds
+    if (timestamp > MILLIS_FROM_EPOCH_THRESHOLD) {
+      return Instant.ofEpochMilli(timestamp).toString();
+    }
+    return Instant.ofEpochSecond(timestamp).toString();
+  }
+
+  public static class CursorField {
+
+    private static final Set<String> STRING_TYPES = Set.of(
+        "STRING", "CHAR", "NCHAR", "NVARCHAR", "VARCHAR", "LONGVARCHAR", "DATE",
+        "TIME", "TIMESTAMP"
+    );
+    private static final Set<String> NUMBER_TYPES = Set.of(
+        "NUMBER", "TINYINT", "SMALLINT", "INT", "INTEGER", "BIGINT", "FLOAT", "DOUBLE",
+        "REAL", "NUMERIC", "DECIMAL"
+    );
+
+    private final String name;
+    private final CursorType type;
+
+    public CursorField(String name, CursorType type) {
+      this.name = name;
+      this.type = type;
+    }
+
+    protected static CursorField fromStream(ConfiguredAirbyteStream stream) {
+      List<String> defaultCursorField = stream.getStream().getDefaultCursorField();
```

Comment:
we should `stream.getCursorField()` instead of `getDefaultCursorField` as some popular source (e.g: mysql or mongo) don't have a default cursor - the user must choose it. The final choice of cursor (whether it's default or configured) always lives in `stream.getCursorField`

Reply:
Fixed!